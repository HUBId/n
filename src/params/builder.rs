use super::types::{
    ChallengeBounds, Endianness, FieldKind, FriFolding, FriParams, HashKind, LdeOrder, LdeParams,
    MerkleArity, MerkleParams, ProofParams, SecurityBudget, TranscriptParams,
};
use super::{ParamsError, StarkParams};

/// Builder used to assemble [`StarkParams`] with validation.
///
/// | Field | Default |
/// |-------|---------|
/// | `params_version` | `1` |
/// | `field` | [`FieldKind::Goldilocks`] |
/// | `hash` | [`HashKind::Poseidon2 { parameter_set: 0 }`] |
/// | `lde` | Blowup `8`, row-major order, coset tag `0x504152414d533031` |
/// | `fri` | `r = 4`, `queries = 30`, `domain_log2 = 22`, natural folding, `num_layers = 5` |
/// | `merkle` | Little-endian leaves, width `4`, binary arity, domain sep `0x4d4b4c5f50415231` |
/// | `transcript` | Protocol tag `0x5354524b5f50524f`, deterministic seed, challenge bounds `16..=64` |
/// | `proof` | Version `1`, `max_size_kb = 512` |
/// | `security` | Target `96` bits, slack `16` bits |
#[derive(Debug, Clone)]
pub struct StarkParamsBuilder {
    pub params_version: u16,
    pub field: FieldKind,
    pub hash: HashKind,
    pub lde: LdeParams,
    pub fri: FriParams,
    pub merkle: MerkleParams,
    pub transcript: TranscriptParams,
    pub proof: ProofParams,
    pub security: SecurityBudget,
}

impl StarkParamsBuilder {
    /// Returns a builder initialised with safe defaults.
    pub fn new() -> Self {
        Self::from_profile(BuiltinProfile::PROFILE_X8)
    }

    /// Loads one of the built-in profiles.
    ///
    /// | Profile | Description | Field | Hash | Blowup | Queries | Arity | Target Bits |
    /// |---------|-------------|-------|------|--------|---------|-------|-------------|
    /// | `PROFILE_X8` | Balanced performance profile with blowup 8 | Goldilocks | Poseidon2 (set 0) | 8 | 30 | Binary | 96 |
    /// | `PROFILE_HISEC_X16` | High security profile with blowup 16 | BN254 | Rescue (set 1) | 16 | 48 | Quaternary | 128 |
    pub fn from_profile(profile: BuiltinProfile) -> Self {
        match profile {
            BuiltinProfile::PROFILE_X8 => StarkParamsBuilder {
                params_version: 1,
                field: FieldKind::Goldilocks,
                hash: HashKind::Poseidon2 { parameter_set: 0 },
                lde: LdeParams {
                    blowup: 8,
                    order: LdeOrder::RowMajor,
                    coset_tag: 0x5041_5241_4d53_3031,
                },
                fri: FriParams {
                    r: 4,
                    queries: 30,
                    domain_log2: 22,
                    folding: FriFolding::Natural,
                    num_layers: 5,
                },
                merkle: MerkleParams {
                    leaf_encoding: Endianness::Little,
                    leaf_width: 4,
                    arity: MerkleArity::Binary,
                    domain_sep: 0x4d4b_4c5f_5041_5231,
                },
                transcript: TranscriptParams {
                    protocol_tag: 0x5354_524b_5f50_524f,
                    seed: *b"RPP-STARK-PROFILE-X8___________0",
                    challenge_bounds: ChallengeBounds {
                        minimum: 16,
                        maximum: 64,
                    },
                },
                proof: ProofParams {
                    version: 1,
                    max_size_kb: 512,
                },
                security: SecurityBudget {
                    target_bits: 96,
                    soundness_slack_bits: 16,
                },
            },
            BuiltinProfile::PROFILE_HISEC_X16 => StarkParamsBuilder {
                params_version: 2,
                field: FieldKind::Bn254,
                hash: HashKind::Rescue { parameter_set: 1 },
                lde: LdeParams {
                    blowup: 16,
                    order: LdeOrder::ColMajor,
                    coset_tag: 0x5041_5241_4d53_4831,
                },
                fri: FriParams {
                    r: 5,
                    queries: 48,
                    domain_log2: 26,
                    folding: FriFolding::Coset,
                    num_layers: 6,
                },
                merkle: MerkleParams {
                    leaf_encoding: Endianness::Big,
                    leaf_width: 8,
                    arity: MerkleArity::Quaternary,
                    domain_sep: 0x484d_524b_5f48_4953,
                },
                transcript: TranscriptParams {
                    protocol_tag: 0x5354_524b_5f48_4953,
                    seed: *b"RPP-STARK-HISEC-X16___________00",
                    challenge_bounds: ChallengeBounds {
                        minimum: 24,
                        maximum: 96,
                    },
                },
                proof: ProofParams {
                    version: 2,
                    max_size_kb: 768,
                },
                security: SecurityBudget {
                    target_bits: 128,
                    soundness_slack_bits: 32,
                },
            },
        }
    }

    /// Validates the builder fields and emits a [`StarkParams`] instance.
    pub fn build(&self) -> Result<StarkParams, ParamsError> {
        StarkParams::try_from_builder(self)
    }
}

/// Supported built-in profiles.
#[allow(non_camel_case_types)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BuiltinProfile {
    /// Balanced performance profile with blowup 8.
    PROFILE_X8,
    /// High security profile with blowup 16.
    PROFILE_HISEC_X16,
}

impl Default for StarkParamsBuilder {
    fn default() -> Self {
        Self::new()
    }
}
