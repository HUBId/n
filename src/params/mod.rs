#![forbid(unsafe_code)]

//! Canonical parameter registry for the STARK pipeline.
//!
//! This module defines [`StarkParams`] as the single source of truth for every
//! security and performance relevant parameter in the proving stack.  The
//! structure is intentionally explicit and designed to be serialised into a
//! deterministic byte layout that is shared between the prover, verifier and
//! tooling such as auditors or snapshot tests.
//!
//! # Overview
//!
//! The parameter space is split into themed sub-structures and enums.  The
//! following table summarises the high level groupings:
//!
//! | Group | Description |
//! |-------|-------------|
//! | Field & Hash | Selection of the prime field and hash function family used throughout the pipeline. |
//! | LDE & FRI | Low Degree Extension and Fast Reed–Solomon IOP knobs governing blowup factors and folding strategies. |
//! | Merkle & Transcript | Commitment encoding rules and Fiat–Shamir transcript framing. |
//! | Proof & Security | Envelope size limits and overall soundness budgeting. |
//!
//! All validation logic lives in [`validate`], canonical serialisation lives in
//! [`ser`] and the stable digest computation is exposed via
//! [`StarkParams::params_hash`].  Consumers are expected to use the
//! [`StarkParamsBuilder`] helper which offers safe defaults and pre-defined
//! [`BuiltinProfile`] presets.
//!
//! # Invariants
//!
//! * The canonical serialisation is strictly ordered as documented in
//!   [`ser`] and is stable across Rust versions.
//! * [`StarkParams::params_hash`] commits to this serialisation and therefore
//!   uniquely identifies compatible parameter sets.
//! * [`StarkParams::is_compatible_with`] only allows variations on
//!   non-critical fields, permitting for example different proof size budgets
//!   while still sharing the same security assumptions.
//!
//! # Examples
//!
//! While the module intentionally avoids executable examples to keep the
//! specification deterministic, the tests and snapshot fixtures demonstrate how
//! `BuiltinProfile::PROFILE_X8` and `BuiltinProfile::PROFILE_HISEC_X16` expand
//! into complete parameter sets.

mod builder;
mod hash;
mod ser;
mod stark_params;
mod types;
mod validate;

pub use crate::ser::{SerError, SerKind};
pub use builder::{BuiltinProfile, StarkParamsBuilder};
pub use hash::params_hash;
pub use ser::{deserialize_params, serialize_params};
pub use stark_params::StarkParams;
pub use types::{
    ChallengeBounds, Endianness, FieldKind, FriFolding, FriParams, HashFamily, HashKind, LdeOrder,
    LdeParams, MerkleArity, MerkleParams, ProofParams, SecurityBudget, TranscriptParams,
};
pub use validate::{ParamsError, ValidationReport};
