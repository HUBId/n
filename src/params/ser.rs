use super::types::{
    ChallengeBounds, Endianness, FieldKind, FriFolding, FriParams, HashFamily, HashKind, LdeOrder,
    LdeParams, MerkleArity, MerkleParams, ProofParams, SecurityBudget, TranscriptParams,
};
use super::StarkParams;

/// Canonical binary serialisation for [`StarkParams`].
///
/// | Offset | Field | Encoding |
/// |--------|-------|----------|
/// | 0..2 | `params_version` | `u16` little-endian |
/// | 2..4 | `field` | `u16` discriminant |
/// | 4..5 | `hash.family` | `u8` discriminant |
/// | 5..7 | `hash.parameter_id` | `u16` little-endian |
/// | 7..11 | `lde.blowup` | `u32` little-endian |
/// | 11..12 | `lde.order` | `u8` discriminant |
/// | 12..20 | `lde.coset_tag` | `u64` little-endian |
/// | 20..21 | `fri.r` | `u8` |
/// | 21..23 | `fri.queries` | `u16` little-endian |
/// | 23..25 | `fri.domain_log2` | `u16` little-endian |
/// | 25..26 | `fri.folding` | `u8` discriminant |
/// | 26..27 | `fri.num_layers` | `u8` |
/// | 27..28 | `merkle.leaf_encoding` | `u8` discriminant |
/// | 28..29 | `merkle.leaf_width` | `u8` |
/// | 29..30 | `merkle.arity` | `u8` value (2 or 4) |
/// | 30..38 | `merkle.domain_sep` | `u64` little-endian |
/// | 38..46 | `transcript.protocol_tag` | `u64` little-endian |
/// | 46..78 | `transcript.seed` | 32 raw bytes |
/// | 78..79 | `transcript.challenge_bounds.minimum` | `u8` |
/// | 79..80 | `transcript.challenge_bounds.maximum` | `u8` |
/// | 80..82 | `proof.version` | `u16` little-endian |
/// | 82..86 | `proof.max_size_kb` | `u32` little-endian |
/// | 86..88 | `security.target_bits` | `u16` little-endian |
/// | 88..89 | `security.soundness_slack_bits` | `u8` |
///
/// The layout intentionally avoids padding and ensures that byte-for-byte
/// equality implies identical parameter sets.
///
/// Serialisation error kind.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SerKind {
    /// Input ended before all fields could be read.
    UnexpectedEnd { field: &'static str },
    /// Encountered an unknown discriminant value.
    InvalidDiscriminant { field: &'static str, value: u16 },
    /// Additional bytes were present after consuming the structure.
    TrailingBytes { expected: usize, remaining: usize },
}

/// Serialises the parameter set into canonical bytes.
pub fn serialize_params(params: &StarkParams) -> Vec<u8> {
    let mut out = Vec::with_capacity(89);
    out.extend_from_slice(&params.params_version().to_le_bytes());
    out.extend_from_slice(&params.field().code().to_le_bytes());
    out.push(params.hash().family().code());
    out.extend_from_slice(&params.hash().parameter_id().to_le_bytes());
    out.extend_from_slice(&params.lde().blowup.to_le_bytes());
    out.push(params.lde().order.code());
    out.extend_from_slice(&params.lde().coset_tag.to_le_bytes());
    out.push(params.fri().r);
    out.extend_from_slice(&params.fri().queries.to_le_bytes());
    out.extend_from_slice(&params.fri().domain_log2.to_le_bytes());
    out.push(params.fri().folding.code());
    out.push(params.fri().num_layers);
    out.push(params.merkle().leaf_encoding.code());
    out.push(params.merkle().leaf_width);
    out.push(params.merkle().arity.code());
    out.extend_from_slice(&params.merkle().domain_sep.to_le_bytes());
    out.extend_from_slice(&params.transcript().protocol_tag.to_le_bytes());
    out.extend_from_slice(&params.transcript().seed);
    out.push(params.transcript().challenge_bounds.minimum);
    out.push(params.transcript().challenge_bounds.maximum);
    out.extend_from_slice(&params.proof().version.to_le_bytes());
    out.extend_from_slice(&params.proof().max_size_kb.to_le_bytes());
    out.extend_from_slice(&params.security().target_bits.to_le_bytes());
    out.push(params.security().soundness_slack_bits);
    out
}

/// Deserialises a parameter set from canonical bytes.
pub fn deserialize_params(bytes: &[u8]) -> Result<StarkParams, SerKind> {
    let mut cursor = Cursor::new(bytes);
    let params_version = cursor.take_u16("params_version")?;
    let field_code = cursor.take_u16("field")?;
    let field = FieldKind::from_code(field_code).ok_or(SerKind::InvalidDiscriminant {
        field: "field",
        value: field_code,
    })?;
    let family_code = cursor.take_u8("hash.family")?;
    let family = HashFamily::from_code(family_code).ok_or(SerKind::InvalidDiscriminant {
        field: "hash.family",
        value: family_code as u16,
    })?;
    let hash_param = cursor.take_u16("hash.parameter_id")?;
    let hash = HashKind::from_codes(family, hash_param);
    let lde_blowup = cursor.take_u32("lde.blowup")?;
    let lde_order_code = cursor.take_u8("lde.order")?;
    let lde_order = LdeOrder::from_code(lde_order_code).ok_or(SerKind::InvalidDiscriminant {
        field: "lde.order",
        value: lde_order_code as u16,
    })?;
    let lde_coset = cursor.take_u64("lde.coset_tag")?;
    let fri_r = cursor.take_u8("fri.r")?;
    let fri_queries = cursor.take_u16("fri.queries")?;
    let fri_domain = cursor.take_u16("fri.domain_log2")?;
    let fri_folding_code = cursor.take_u8("fri.folding")?;
    let fri_folding =
        FriFolding::from_code(fri_folding_code).ok_or(SerKind::InvalidDiscriminant {
            field: "fri.folding",
            value: fri_folding_code as u16,
        })?;
    let fri_layers = cursor.take_u8("fri.num_layers")?;
    let merkle_encoding_code = cursor.take_u8("merkle.leaf_encoding")?;
    let merkle_encoding =
        Endianness::from_code(merkle_encoding_code).ok_or(SerKind::InvalidDiscriminant {
            field: "merkle.leaf_encoding",
            value: merkle_encoding_code as u16,
        })?;
    let merkle_leaf_width = cursor.take_u8("merkle.leaf_width")?;
    let merkle_arity_code = cursor.take_u8("merkle.arity")?;
    let merkle_arity =
        MerkleArity::from_code(merkle_arity_code).ok_or(SerKind::InvalidDiscriminant {
            field: "merkle.arity",
            value: merkle_arity_code as u16,
        })?;
    let merkle_domain_sep = cursor.take_u64("merkle.domain_sep")?;
    let transcript_protocol = cursor.take_u64("transcript.protocol_tag")?;
    let transcript_seed = cursor.take_array::<32>("transcript.seed")?;
    let transcript_ch_min = cursor.take_u8("transcript.challenge_bounds.minimum")?;
    let transcript_ch_max = cursor.take_u8("transcript.challenge_bounds.maximum")?;
    let proof_version = cursor.take_u16("proof.version")?;
    let proof_max_size = cursor.take_u32("proof.max_size_kb")?;
    let security_target = cursor.take_u16("security.target_bits")?;
    let security_slack = cursor.take_u8("security.soundness_slack_bits")?;

    if cursor.remaining() != 0 {
        return Err(SerKind::TrailingBytes {
            expected: cursor.position,
            remaining: cursor.remaining(),
        });
    }

    Ok(StarkParams {
        params_version,
        field,
        hash,
        lde: LdeParams {
            blowup: lde_blowup,
            order: lde_order,
            coset_tag: lde_coset,
        },
        fri: FriParams {
            r: fri_r,
            queries: fri_queries,
            domain_log2: fri_domain,
            folding: fri_folding,
            num_layers: fri_layers,
        },
        merkle: MerkleParams {
            leaf_encoding: merkle_encoding,
            leaf_width: merkle_leaf_width,
            arity: merkle_arity,
            domain_sep: merkle_domain_sep,
        },
        transcript: TranscriptParams {
            protocol_tag: transcript_protocol,
            seed: transcript_seed,
            challenge_bounds: ChallengeBounds {
                minimum: transcript_ch_min,
                maximum: transcript_ch_max,
            },
        },
        proof: ProofParams {
            version: proof_version,
            max_size_kb: proof_max_size,
        },
        security: SecurityBudget {
            target_bits: security_target,
            soundness_slack_bits: security_slack,
        },
    })
}

struct Cursor<'a> {
    bytes: &'a [u8],
    position: usize,
}

impl<'a> Cursor<'a> {
    fn new(bytes: &'a [u8]) -> Self {
        Self { bytes, position: 0 }
    }

    fn take_u8(&mut self, field: &'static str) -> Result<u8, SerKind> {
        if self.position + 1 > self.bytes.len() {
            return Err(SerKind::UnexpectedEnd { field });
        }
        let value = self.bytes[self.position];
        self.position += 1;
        Ok(value)
    }

    fn take_u16(&mut self, field: &'static str) -> Result<u16, SerKind> {
        if self.position + 2 > self.bytes.len() {
            return Err(SerKind::UnexpectedEnd { field });
        }
        let value = u16::from_le_bytes([self.bytes[self.position], self.bytes[self.position + 1]]);
        self.position += 2;
        Ok(value)
    }

    fn take_u32(&mut self, field: &'static str) -> Result<u32, SerKind> {
        if self.position + 4 > self.bytes.len() {
            return Err(SerKind::UnexpectedEnd { field });
        }
        let value = u32::from_le_bytes([
            self.bytes[self.position],
            self.bytes[self.position + 1],
            self.bytes[self.position + 2],
            self.bytes[self.position + 3],
        ]);
        self.position += 4;
        Ok(value)
    }

    fn take_u64(&mut self, field: &'static str) -> Result<u64, SerKind> {
        if self.position + 8 > self.bytes.len() {
            return Err(SerKind::UnexpectedEnd { field });
        }
        let value = u64::from_le_bytes([
            self.bytes[self.position],
            self.bytes[self.position + 1],
            self.bytes[self.position + 2],
            self.bytes[self.position + 3],
            self.bytes[self.position + 4],
            self.bytes[self.position + 5],
            self.bytes[self.position + 6],
            self.bytes[self.position + 7],
        ]);
        self.position += 8;
        Ok(value)
    }

    fn take_array<const N: usize>(&mut self, field: &'static str) -> Result<[u8; N], SerKind> {
        if self.position + N > self.bytes.len() {
            return Err(SerKind::UnexpectedEnd { field });
        }
        let mut out = [0u8; N];
        out.copy_from_slice(&self.bytes[self.position..self.position + N]);
        self.position += N;
        Ok(out)
    }

    fn remaining(&self) -> usize {
        self.bytes.len().saturating_sub(self.position)
    }
}
