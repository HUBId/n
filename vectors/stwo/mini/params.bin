0100010003200008000000013130534d415241500440000c000105010102315241505f4c4b4d4f52505f4b5254535250502d535441524b2d50524f46494c452d58385f5f5f5f5f5f5f5f5f5f5f301040010000020000600010
