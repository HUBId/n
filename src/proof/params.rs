use crate::config::{
    ProfileConfig, MERKLE_SCHEME_ID_BLAKE3_4ARY_V1, PROFILE_HIGH_SECURITY_CONFIG,
    PROFILE_STANDARD_CONFIG,
};
use crate::params::{FriFolding, HashKind, LdeOrder, MerkleArity, StarkParams, StarkParamsBuilder};

/// Returns the canonical Stark parameter set shared by prover and verifier.
///
/// The current implementation relies on the deterministic profile used by the
/// built-in prover.  Both parties fix the hashing backend to Blake2s and commit
/// to single-field-element Merkle leaves.  The helper mirrors the prover
/// configuration so callers can rely on identical Merkle framing without
/// leaking the builder details elsewhere in the crate.
pub fn canonical_stark_params(profile: &ProfileConfig) -> StarkParams {
    let mut builder = StarkParamsBuilder::new();
    builder.hash = HashKind::Blake2s { digest_size: 32 };
    builder.merkle.leaf_width = 1;
    if profile.merkle_scheme_id == MERKLE_SCHEME_ID_BLAKE3_4ARY_V1 {
        builder.merkle.arity = MerkleArity::Quaternary;
    } else {
        builder.merkle.arity = MerkleArity::Binary;
    }
    builder.lde.blowup = profile.lde_factor as u32;
    builder.fri.queries = profile.fri_queries;
    builder.fri.domain_log2 = profile.fri_depth_range.max as u16;
    builder.fri.num_layers = profile
        .fri_depth_range
        .max
        .saturating_sub(profile.fri_depth_range.min)
        .saturating_add(1);

    if profile.id == PROFILE_HIGH_SECURITY_CONFIG.id {
        builder.lde.order = LdeOrder::ColMajor;
        builder.fri.folding = FriFolding::Coset;
    } else if profile.id == PROFILE_STANDARD_CONFIG.id {
        builder.lde.order = LdeOrder::RowMajor;
        builder.fri.folding = FriFolding::Natural;
    }

    builder
        .build()
        .expect("canonical Stark parameters must be valid")
}
