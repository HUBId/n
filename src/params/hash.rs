use crate::hash::config::BLAKE2S_PARAMETERS_V1;
use crate::hash::deterministic::hash;

use super::ser::serialize_params;
use super::types::HashFamily;
use super::StarkParams;

/// Computes the canonical parameter digest for a [`StarkParams`] instance.
///
/// The function reuses the selected proof hash family for the meta commitment:
/// if the proof profile relies on Poseidon2 or Rescue the same algebraic
/// family is reused with the parameter identifier fixed to `0`.  Otherwise a
/// deterministic Blake2s-style sponge is used.  This keeps the parameter digest
/// aligned with the commitment scheme while remaining deterministic.
pub fn params_hash(params: &StarkParams) -> [u8; 32] {
    let payload = serialize_params(params);
    let mut prefixed = Vec::with_capacity(payload.len() + 8);
    let family = params.hash().family();
    match family {
        HashFamily::Poseidon2 => prefixed.extend_from_slice(b"POSEIDON2/0"),
        HashFamily::Rescue => prefixed.extend_from_slice(b"RESCUE/0"),
        HashFamily::Blake2s => prefixed.extend_from_slice(BLAKE2S_PARAMETERS_V1.domain_tag),
    }
    prefixed.extend_from_slice(&payload);
    hash(&prefixed).into()
}
