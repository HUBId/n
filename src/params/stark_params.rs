use serde::{Deserialize, Serialize};

use super::hash::params_hash;
use super::types::{
    FieldKind, FriParams, HashKind, LdeParams, MerkleParams, ProofParams, SecurityBudget,
    TranscriptParams,
};
use super::validate::ParamsError;

/// Canonical STARK parameter set shared by prover and verifier.
///
/// | Field | Type | Description |
/// |-------|------|-------------|
/// | `params_version` | `u16` | Version of the parameter schema. |
/// | `field` | [`FieldKind`] | Prime field used for polynomial arithmetic. |
/// | `hash` | [`HashKind`] | Hash function selection including parameter identifier. |
/// | `lde` | [`LdeParams`] | Low Degree Extension configuration. |
/// | `fri` | [`FriParams`] | Fast Reed–Solomon IOP configuration. |
/// | `merkle` | [`MerkleParams`] | Merkle commitment encoding options. |
/// | `transcript` | [`TranscriptParams`] | Fiat–Shamir transcript framing. |
/// | `proof` | [`ProofParams`] | Proof envelope layout. |
/// | `security` | [`SecurityBudget`] | Global soundness budget. |
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct StarkParams {
    pub(crate) params_version: u16,
    pub(crate) field: FieldKind,
    pub(crate) hash: HashKind,
    pub(crate) lde: LdeParams,
    pub(crate) fri: FriParams,
    pub(crate) merkle: MerkleParams,
    pub(crate) transcript: TranscriptParams,
    pub(crate) proof: ProofParams,
    pub(crate) security: SecurityBudget,
}

impl StarkParams {
    /// Returns the parameter schema version.
    pub const fn params_version(&self) -> u16 {
        self.params_version
    }

    /// Returns the selected prime field.
    pub const fn field(&self) -> FieldKind {
        self.field
    }

    /// Returns the configured hash function.
    pub const fn hash(&self) -> HashKind {
        self.hash
    }

    /// Returns the Low Degree Extension configuration.
    pub const fn lde(&self) -> &LdeParams {
        &self.lde
    }

    /// Returns the FRI configuration.
    pub const fn fri(&self) -> &FriParams {
        &self.fri
    }

    /// Returns the Merkle configuration.
    pub const fn merkle(&self) -> &MerkleParams {
        &self.merkle
    }

    /// Returns the transcript configuration.
    pub const fn transcript(&self) -> &TranscriptParams {
        &self.transcript
    }

    /// Returns the proof envelope configuration.
    pub const fn proof(&self) -> &ProofParams {
        &self.proof
    }

    /// Returns the security budget configuration.
    pub const fn security(&self) -> &SecurityBudget {
        &self.security
    }

    /// Computes the canonical parameter hash.
    ///
    /// The digest is computed over the canonical byte layout defined in
    /// [`crate::params::serialize_params`].
    pub fn params_hash(&self) -> [u8; 32] {
        params_hash(self)
    }

    /// Produces a human-readable profile identifier.
    ///
    /// The identifier is deterministic and contains only ASCII alphanumeric
    /// characters and underscores.  It does not leak any secret material.
    pub fn profile_id(&self) -> String {
        format!(
            "{}_F{}_H{}{}_B{}_Q{}_A{}_V{}",
            match self.security.target_bits {
                bits if bits >= 128 => "PROFILE_HISEC",
                _ => "PROFILE",
            },
            self.field.code(),
            self.hash.family().code(),
            self.hash.parameter_id(),
            self.lde.blowup,
            self.fri.queries,
            self.merkle.arity.code(),
            self.proof.version
        )
    }

    /// Checks whether two parameter sets are compatible on security-critical fields.
    pub fn is_compatible_with(&self, other: &Self) -> bool {
        self.field == other.field
            && self.hash == other.hash
            && self.merkle.arity == other.merkle.arity
            && self.merkle.leaf_encoding == other.merkle.leaf_encoding
            && self.merkle.leaf_width == other.merkle.leaf_width
            && self.transcript.protocol_tag == other.transcript.protocol_tag
            && self.proof.version == other.proof.version
    }

    pub(crate) fn try_from_builder(
        builder: &super::builder::StarkParamsBuilder,
    ) -> Result<Self, ParamsError> {
        let params = Self {
            params_version: builder.params_version,
            field: builder.field,
            hash: builder.hash,
            lde: builder.lde,
            fri: builder.fri,
            merkle: builder.merkle,
            transcript: builder.transcript,
            proof: builder.proof,
            security: builder.security,
        };
        let _ = super::validate::validate(&params)?;
        Ok(params)
    }
}
