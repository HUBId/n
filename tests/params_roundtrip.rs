use proptest::prelude::*;
use rpp_stark::params::{
    deserialize_params, params_hash, serialize_params, BuiltinProfile, FieldKind, HashKind,
    ParamsError, StarkParams, StarkParamsBuilder,
};

fn arb_params() -> impl Strategy<Value = StarkParams> {
    prop_oneof![
        Just(StarkParamsBuilder::new()),
        Just(StarkParamsBuilder::from_profile(
            BuiltinProfile::PROFILE_HISEC_X16
        )),
    ]
    .prop_flat_map(|base| {
        (
            0u32..=4u32,
            0u16..=32u16,
            0u16..=8u16,
            0u8..=4u8,
            0u8..=8u8,
            0u32..=512u32,
            64u16..=256u16,
            0u8..=64u8,
            prop_oneof![Just(false), Just(true)],
        )
            .prop_map(
                move |(
                    extra_blowup,
                    extra_queries,
                    extra_domain,
                    extra_layers,
                    extra_leaf,
                    extra_size,
                    target_bits,
                    slack_bits,
                    flip_hash,
                )| {
                    let mut builder = base.clone();
                    builder.lde.blowup = (builder.lde.blowup + extra_blowup).max(2);
                    builder.fri.queries = (builder.fri.queries + extra_queries).max(1);
                    builder.fri.domain_log2 = (builder.fri.domain_log2 + extra_domain).max(8);
                    builder.fri.num_layers = (builder.fri.num_layers + extra_layers).max(1);
                    builder.merkle.leaf_width = (builder.merkle.leaf_width + extra_leaf).max(1);
                    builder.proof.max_size_kb = builder.proof.max_size_kb + extra_size;
                    builder.security.target_bits = target_bits;
                    let slack_limit = (builder.security.target_bits / 2) as u8;
                    builder.security.soundness_slack_bits = slack_bits.min(slack_limit);
                    if flip_hash {
                        builder.hash = HashKind::Blake2s { digest_size: 32 };
                    } else {
                        builder.hash = match builder.field {
                            FieldKind::Goldilocks => HashKind::Poseidon2 { parameter_set: 0 },
                            FieldKind::Bn254 => HashKind::Rescue { parameter_set: 1 },
                        };
                    }
                    builder.build().expect("valid randomized params")
                },
            )
    })
}

#[test]
fn canonical_roundtrip() {
    let params = StarkParamsBuilder::new().build().expect("valid");
    let bytes = serialize_params(&params);
    let decoded = deserialize_params(&bytes).expect("deserialise");
    assert_eq!(params, decoded);
}

#[test]
fn bincode_roundtrip() {
    let params = StarkParamsBuilder::from_profile(BuiltinProfile::PROFILE_HISEC_X16)
        .build()
        .expect("valid profile");
    let bytes = bincode::serialize(&params).expect("serialize");
    let decoded: StarkParams = bincode::deserialize(&bytes).expect("deserialize");
    assert_eq!(params, decoded);
}

#[test]
fn compatibility_relaxes_non_critical() {
    let mut builder = StarkParamsBuilder::new();
    let base = builder.build().unwrap();
    builder.proof.max_size_kb += 128;
    let tweaked = builder.build().unwrap();
    assert!(base.is_compatible_with(&tweaked));
}

#[test]
fn compatibility_rejects_hash_change() {
    let base = StarkParamsBuilder::new().build().unwrap();
    let mut builder = StarkParamsBuilder::new();
    builder.hash = HashKind::Blake2s { digest_size: 32 };
    let altered = builder.build().unwrap();
    assert!(!base.is_compatible_with(&altered));
}

#[test]
fn invalid_blowup() {
    let mut builder = StarkParamsBuilder::new();
    builder.lde.blowup = 1;
    let err = builder.build().unwrap_err();
    assert!(matches!(err, ParamsError::InvalidBlowup { min: 2, got: 1 }));
}

#[test]
fn invalid_queries() {
    let mut builder = StarkParamsBuilder::new();
    builder.fri.queries = 0;
    let err = builder.build().unwrap_err();
    assert!(matches!(
        err,
        ParamsError::InvalidQueries { min: 1, got: 0 }
    ));
}

#[test]
fn invalid_leaf_width() {
    let mut builder = StarkParamsBuilder::new();
    builder.merkle.leaf_width = 0;
    let err = builder.build().unwrap_err();
    assert!(matches!(err, ParamsError::LeafWidthZero));
}

#[test]
fn version_mismatch() {
    let mut builder = StarkParamsBuilder::new();
    builder.proof.version = builder.params_version + 1;
    let err = builder.build().unwrap_err();
    assert!(matches!(err, ParamsError::VersionMismatch { .. }));
}

#[test]
fn domain_log2_too_small() {
    let mut builder = StarkParamsBuilder::new();
    builder.fri.domain_log2 = 4;
    let err = builder.build().unwrap_err();
    assert!(matches!(err, ParamsError::DomainTooSmall { .. }));
}

#[test]
fn params_hash_stable_for_profile() {
    let params = StarkParamsBuilder::new().build().unwrap();
    let hash = params.params_hash();
    let bytes = serialize_params(&params);
    let decoded = deserialize_params(&bytes).unwrap();
    assert_eq!(hash, decoded.params_hash());
}

proptest! {
    #[test]
    fn prop_roundtrip_idempotent(params in arb_params()) {
        let bytes = serialize_params(&params);
        let decoded = deserialize_params(&bytes).unwrap();
        let decoded_clone = decoded.clone();
        prop_assert_eq!(params, decoded_clone);
        let bytes_again = serialize_params(&decoded);
        prop_assert_eq!(bytes, bytes_again);
    }

    #[test]
    fn prop_hash_deterministic(params in arb_params()) {
        let hash1 = params_hash(&params);
        let hash2 = params_hash(&params);
        prop_assert_eq!(hash1, hash2);
    }
}

#[test]
#[ignore]
fn snapshot_profiles() {
    let profile_x8 = StarkParamsBuilder::new().build().unwrap();
    let hisec = StarkParamsBuilder::from_profile(BuiltinProfile::PROFILE_HISEC_X16)
        .build()
        .unwrap();
    insta::assert_json_snapshot!("profile_x8_params", &profile_x8);
    insta::assert_snapshot!(
        "profile_x8_hash",
        format!("{:02x?}", profile_x8.params_hash())
    );
    insta::assert_json_snapshot!("profile_hisec_params", &hisec);
    insta::assert_snapshot!(
        "profile_hisec_hash",
        format!("{:02x?}", hisec.params_hash())
    );
}
