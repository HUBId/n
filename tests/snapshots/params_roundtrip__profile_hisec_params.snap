---
source: tests/params_roundtrip.rs
expression: "&hisec"
---
{
  "params_version": 2,
  "field": "Bn254",
  "hash": {
    "Rescue": {
      "parameter_set": 1
    }
  },
  "lde": {
    "blowup": 16,
    "order": "ColMajor",
    "coset_tag": 5782993836944082993
  },
  "fri": {
    "r": 5,
    "queries": 48,
    "domain_log2": 26,
    "folding": "Coset",
    "num_layers": 6
  },
  "merkle": {
    "leaf_encoding": "Big",
    "leaf_width": 8,
    "arity": "Quaternary",
    "domain_sep": 5209910827612129619
  },
  "transcript": {
    "protocol_tag": 6004514686866311507,
    "seed": [
      82,
      80,
      80,
      45,
      83,
      84,
      65,
      82,
      75,
      45,
      72,
      73,
      83,
      69,
      67,
      45,
      88,
      49,
      54,
      95,
      95,
      95,
      95,
      95,
      95,
      95,
      95,
      95,
      95,
      95,
      48,
      48
    ],
    "challenge_bounds": {
      "minimum": 24,
      "maximum": 96
    }
  },
  "proof": {
    "version": 2,
    "max_size_kb": 768
  },
  "security": {
    "target_bits": 128,
    "soundness_slack_bits": 32
  }
}
